File: services/transaction-service/activemq.py
import os
import json
import stomp
import time
import threading
import logging
from typing import Callable, Optional

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [%(levelname)s] %(message)s',
    handlers=[logging.FileHandler("transaction-service.log"), logging.StreamHandler()]
)
logger = logging.getLogger(__name__)

# Load environment variables
BROKER_HOST = os.getenv("ACTIVEMQ_BROKER", "shared-activemq")
BROKER_PORT = int(os.getenv("ACTIVEMQ_BROKER_PORT", "61616"))
USERNAME = os.getenv("ACTIVEMQ_USERNAME", "admin")
PASSWORD = os.getenv("ACTIVEMQ_PASSWORD", "admin")

# Message broker configuration
class MessageListener(stomp.ConnectionListener):
    def __init__(self, callback: Optional[Callable[[str], None]] = None):
        self.callback = callback

    def on_error(self, frame):
        logger.error(f"ActiveMQ error: {frame.body}")

    def on_message(self, frame):
        body = frame.body
        logger.info(f"Received message: {body}")
        if self.callback:
            try:
                self.callback(body)
            except Exception as e:
                logger.error(f"Error processing message: {str(e)}")
        frame.ack()

    def on_disconnected(self):
        logger.warning("Disconnected from ActiveMQ")

    def on_connected(self, frame):
        logger.info("Connected to ActiveMQ")

# Connection manager with reconnect logic
class ActiveMQConnection:
    def __init__(self, host: str, port: int, username: str, password: str):
        self.host = host
        self.port = port
        self.username = username
        self.password = password
        self.conn = None
        self.lock = threading.Lock()

    def connect(self):
        with self.lock:
            if not self.conn or not self.conn.is_connected():
                try:
                    self.conn = stomp.Connection([(self.host, self.port)])
                    self.conn.connect(self.username, self.password, wait=True)
                    logger.info(f"Connected to ActiveMQ at {self.host}:{self.port}")
                except Exception as e:
                    logger.error(f"Failed to connect to ActiveMQ: {str(e)}")
                    raise

    def disconnect(self):
        with self.lock:
            if self.conn and self.conn.is_connected():
                self.conn.disconnect()
                logger.info("Disconnected from ActiveMQ")

    def send_message(self, destination: str, message: str):
        self.connect()
        try:
            self.conn.send(body=message, destination=destination)
            logger.info(f"Sent to {destination}: {message}")
        except Exception as e:
            logger.error(f"Failed to send message to {destination}: {str(e)}")
            self.disconnect()
            raise

    def subscribe(self, destination: str, listener: MessageListener):
        self.connect()
        try:
            self.conn.set_listener("", listener)
            self.conn.subscribe(destination=destination, id=1, ack="client")
            logger.info(f"Subscribed to {destination}")
        except Exception as e:
            logger.error(f"Failed to subscribe to {destination}: {str(e)}")
            self.disconnect()
            raise

# Global connection instance
mq_conn = ActiveMQConnection(BROKER_HOST, BROKER_PORT, USERNAME, PASSWORD)

# Function to send a book purchase event
def send_book_purchase_event(user_id: str, book_id: int, transaction_type: str):
    message = json.dumps({
        "event": "book_purchase",
        "userId": user_id,
        "bookId": book_id,
        "transactionType": transaction_type,
        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
    })
    try:
        mq_conn.send_message("/topic/book-purchases", message)
    except Exception as e:
        logger.error(f"Failed to send book purchase event: {str(e)}")

# Function to subscribe to a topic/queue
def subscribe_to_topic(topic: str, callback: Optional[Callable[[str], None]] = None):
    def run_subscription():
        while True:
            try:
                listener = MessageListener(callback)
                mq_conn.subscribe(topic, listener)
                # Keep connection alive
                while mq_conn.conn and mq_conn.conn.is_connected():
                    time.sleep(1)
            except Exception as e:
                logger.error(f"Subscription error for {topic}: {str(e)}")
                time.sleep(5)  # Wait before reconnecting

    threading.Thread(target=run_subscription, daemon=True).start()

# Example callback for processing book purchase messages
def handle_book_purchase_message(message: str):
    try:
        data = json.loads(message)
        if data.get("event") == "book_purchase":
            logger.info(f"Processed book purchase: userId={data['userId']}, bookId={data['bookId']}")
            # Add logic to update transaction state if needed
    except json.JSONDecodeError:
        logger.error(f"Invalid JSON message: {message}")

# Subscribe to book purchases
subscribe_to_topic("/topic/book-purchases", handle_book_purchase_message)

if __name__ == "__main__":
    logger.info("Transaction service ActiveMQ setup complete")

File: services/book-service/src/activemq.js
const stompit = require('stompit');
const express = require('express');
const app = express();
app.use(express.json());

// Load environment variables
const BROKER_URL = process.env.ACTIVEMQ_BROKER_URL || 'tcp://shared-activemq:61616';
const USERNAME = process.env.ACTIVEMQ_USERNAME || 'admin';
const PASSWORD = process.env.ACTIVEMQ_PASSWORD || 'admin';
const PORT = process.env.PORT || 3001; // Changed from 50051 to avoid gRPC conflict

// Parse brokerUrl
const url = new URL(BROKER_URL.replace('tcp://', 'http://'));
const BROKER_HOST = url.hostname;
const BROKER_PORT = parseInt(url.port) || 61616;

// Configure reconnect options
const reconnectOptions = {
    maxReconnects: 10,
    reconnectDelay: 5000,
    connectOptions: {
        host: BROKER_HOST,
        port: BROKER_PORT,
        connectHeaders: {
            'host': '/',
            'login': USERNAME,
            'passcode': PASSWORD,
            'heart-beat': '5000,5000'
        }
    }
};

// Function to send a message
function sendMessage(destination, message) {
    stompit.connect(reconnectOptions.connectOptions, (error, client) => {
        if (error) {
            console.error(`Failed to connect to ActiveMQ: ${error.message}`);
            return;
        }
        const frame = client.send({ destination });
        frame.write(JSON.stringify(message));
        frame.end();
        client.disconnect();
        console.log(`Sent to ${destination}: ${JSON.stringify(message)}`);
    });
}

// Function to subscribe to a topic/queue
function subscribeToTopic(topic, callback) {
    const channel = new stompit.Channel(reconnectOptions);
    channel.subscribe({ destination: topic, ack: 'client' }, (error, message) => {
        if (error) {
            console.error(`Subscribe failed for ${topic}: ${error.message}`);
            return;
        }
        message.readString('utf-8', (error, body) => {
            if (error) {
                console.error(`Read message failed: ${error.message}`);
                return;
            }
            console.log(`Received from ${topic}: ${body}`);
            try {
                callback(JSON.parse(body));
                message.ack();
            } catch (e) {
                console.error(`Error processing message: ${e.message}`);
            }
        });
    });
}

// Health check endpoint
app.get('/health', (req, res) => {
    res.status(200).send('Server is up');
});

// Send message endpoint (for testing)
app.post('/send-purchase', (req, res) => {
    const { userId, bookId, transactionType } = req.body;
    const message = {
        event: 'book_purchase',
        userId,
        bookId,
        transactionType,
        timestamp: new Date().toISOString()
    };
    sendMessage('/topic/book-purchases', message);
    res.send('Book purchase message sent');
});

// Handle book purchase messages
subscribeToTopic('/topic/book-purchases', (message) => {
    if (message.event === 'book_purchase') {
        console.log(`Book purchased: userId=${message.userId}, bookId=${message.bookId}`);
        // Add logic to update book availability
    }
});

app.listen(PORT, () => {
    console.log(`Book service running on port ${PORT}`);
});